// @sap/cf-tools/out/src/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.eOrderDirection = exports.eOperation = exports.eFilters = exports.CF_CMD_EXIT_CODE = exports.DEFAULT_TARGET = exports.CF_PAGE_SIZE = exports.NEW_LINE = exports.OK = void 0;
exports.OK = "OK";
exports.NEW_LINE = "\n";
exports.CF_PAGE_SIZE = 99 * 3;
exports.DEFAULT_TARGET = "Default (no targets)";
var CF_CMD_EXIT_CODE;
(function (CF_CMD_EXIT_CODE) {
    CF_CMD_EXIT_CODE[CF_CMD_EXIT_CODE["OK"] = 0] = "OK";
    CF_CMD_EXIT_CODE[CF_CMD_EXIT_CODE["ERROR"] = -1] = "ERROR";
    CF_CMD_EXIT_CODE[CF_CMD_EXIT_CODE["CANCEL_REQ"] = -2] = "CANCEL_REQ";
    CF_CMD_EXIT_CODE[CF_CMD_EXIT_CODE["CANCELED"] = -3] = "CANCELED";
})(CF_CMD_EXIT_CODE = exports.CF_CMD_EXIT_CODE || (exports.CF_CMD_EXIT_CODE = {}));
var eFilters;
(function (eFilters) {
    eFilters["name"] = "name";
    eFilters["space_guid"] = "space_guid";
    eFilters["service_plan_guid"] = "service_plan_guid";
    eFilters["service_binding_guid"] = "service_binding_guid";
    eFilters["gateway_name"] = "gateway_name";
    eFilters["organization_guid"] = "organization_guid";
    eFilters["service_key_guid"] = "service_key_guid";
    eFilters["service_guid"] = "service_guid";
    eFilters["service_instance_guid"] = "service_instance_guid";
    eFilters["user_guid"] = "user_guid";
    eFilters["manager_guid"] = "manager_guid";
    eFilters["billing_manager_guid"] = "billing_manager_guid";
    eFilters["auditor_guid"] = "auditor_guid";
    eFilters["status"] = "status";
    eFilters["label"] = "label";
})(eFilters = exports.eFilters || (exports.eFilters = {}));
var eOperation;
(function (eOperation) {
    eOperation["eq"] = ":";
    eOperation["be"] = ">=";
    eOperation["le"] = "<=";
    eOperation["l"] = "<";
    eOperation["b"] = ">";
    eOperation["IN"] = "%20IN%20";
})(eOperation = exports.eOperation || (exports.eOperation = {}));
var eOrderDirection;
(function (eOrderDirection) {
    eOrderDirection[eOrderDirection["asc"] = 0] = "asc";
    eOrderDirection[eOrderDirection["desc"] = 1] = "desc";
})(eOrderDirection = exports.eOrderDirection || (exports.eOrderDirection = {}));
